File: app/src/main/java/com/dingyi/visualizer/drawable/LineDrawable.java
package com.dingyi.visualizer.drawable;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;

public class LineDrawable extends BaseDrawable {

    private final Paint linePaint=new Paint();

    private float width;

    private float height;

    private float lineCount;

    public LineDrawable(Context context) {
        super(context);
    }

    @Override
    public void init() {
        linePaint.setStrokeWidth(2);
        linePaint.setColor(getDataBean().getColorData().getLinePaintColor());
    }


    private void readyData(Canvas canvas) {
       width=canvas.getWidth();
       height=canvas.getHeight();
       lineCount= width/ getDataBean().getLineWidth();
    }

    @Override
    public void onDraw(Canvas canvas) {

        if (lineCount == 0) {
            readyData(canvas);
        }

        for (int i=1;i<lineCount;i++) {
            float des = (float) getData()[8 + i];
            canvas.drawRect((i-1)*getDataBean().getLineWidth(),height-((height/3)/120*des),
                    i*getDataBean().getLineWidth()-getDataBean().getLineMargin(),height,linePaint);
        }

        System.gc();

    }
}

File: app/src/main/kotlin/io/github/dingyi222666/visualizer_app/drawable/LineDrawable.kt
package io.github.dingyi222666.visualizer_app.drawable

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.RectF
import io.github.dingyi222666.visualizer_app.visualizer.VisualizerConfig

/**
 * Line-based audio visualization drawable
 */
class LineDrawable(context: Context) : BaseDrawable(context) {
    
    // Paint for drawing lines
    private val linePaint = Paint()
    
    // Canvas dimensions
    private var width: Float = 0f
    private var height: Float = 0f
    
    // Number of lines to draw
    private var lineCount: Int = 0
    
    // Reusable rectangle for drawing
    private val rect = RectF()
    
    /**
     * Initialize the drawable with default settings
     */
    override fun init() {
        val config = getConfig() ?: return
        
        linePaint.apply {
            strokeWidth = 2f
            color = config.colors.primaryColor
            style = Paint.Style.FILL
        }
    }
    
    /**
     * Called when configuration changes
     */
    override fun onConfigChanged(config: VisualizerConfig) {
        init()
    }
    
    /**
     * Initialize dimensions based on canvas size
     */
    private fun prepareDrawing(canvas: Canvas) {
        val config = getConfig() ?: return
        
        width = canvas.width.toFloat()
        height = canvas.height.toFloat()
        lineCount = (width / config.line.width).toInt()
    }
    
    /**
     * Draw the visualization
     */
    override fun onDraw(canvas: Canvas) {
        val config = getConfig() ?: return
        val data = getData() ?: return
        
        // Initialize dimensions if needed
        if (lineCount == 0) {
            prepareDrawing(canvas)
        }
        
        val startIndex = config.dataStartIndex
        val lineWidth = config.line.width
        val lineMargin = config.line.margin
        
        // Draw each bar
        for (i in 1 until lineCount) {
            val dataIndex = startIndex + i
            
            // Get amplitude for this bar (default to 1 if out of bounds)
            val amplitude = if (dataIndex < data.size) data[dataIndex].toFloat() else 1f
            
            // Scale the amplitude to the display height
            val barHeight = (height / 3) / 120 * amplitude
            
            // Calculate bar rectangle
            rect.left = (i - 1) * lineWidth
            rect.top = height - barHeight
            rect.right = i * lineWidth - lineMargin
            rect.bottom = height
            
            // Draw the bar
            canvas.drawRect(rect, linePaint)
        }
    }
} 